src/Symfony/Component/DependencyInjection/Definition.php

```diff
@@ -694,28 +695,73 @@ public function getDeprecation(string $id): array
      * @param string|array|Reference|null $configurator A PHP function, reference or an array containing a class/Reference and a method to call
      *
      * @return $this
+     * @deprecated since Symfony 7.1 and will be removed in 8.0, use addConfigurator() or setConfigurators() instead
      */
     public function setConfigurator(string|array|Reference|null $configurator): static
     {
-        $this->changes['configurator'] = true;
+        trigger_deprecation('symfony/dependency-injection', '7.1', 'Method "%s()" is deprecated, use "%s()" or "%s()" instead.', __METHOD__, 'setConfigurators', 'addConfigurator');
+        return $this->setConfigurators([$configurator]);
+    }
+
+    /**
+     * Add a configurator to the configurator list.
+     *
+     * @param string|array|Reference|null $configurator A PHP function, reference or an array containing a class/Reference and a method to call
```

Comment:
`null` isn't valid. It was used to unset the configurator.

You can describe the array shape:
```suggestion
     * @param callable|Reference|array{0:class-string|Reference,1:string} $configurators A PHP function, reference or an array containing a class/Reference and a method to call
```